### PR TITLE
Implement --use-native-builtins flag.

### PR DESCRIPTION
**Pull request by aarongreig**
This allows avoiding builtin library implementations of specified builtin functions.

Also add a handful of tests for the new flag, builtins selected are all functions that we rely on libclc for, and that map directly to a glsl instruction.

This is the most generic version of the flag with the most validation, I had to shuffle stuff around to get the lookups working and it fails noisily if you pass it a builtin that we can't replace. I could revert some of the validation related changes if we restricted it to only functions we can map directly to a glsl extended instruction and it would still serve my purposes just fine, but I figured the maximum utility version was the best first proposal.

Based on discussion in #888 

This contribution is being made by Codeplay on behalf of Samsung.

### CODE REVIEW

**Comment by aarongreig** [comment]
CI seems to have an issue finding the vulkan headers
```
/bin/clang-8.0.0/bin/clang++  -DCOMPILER_AVAILABLE -I../external/OpenCL-Headers -I/tmpfs/src/github/clspv/third_party/llvm/llvm/include -Iexternal/clspv/third_party/llvm/include -I/tmpfs/src/github/clspv/third_party/llvm/llvm/utils/unittest/googletest/include -I../tests/../src -I../external/SPIRV-Tools/include -I../external/SPIRV-Headers/include -I/tmpfs/src/github/clspv/include -Iexternal/clspv/include -I/tmpfs/src/github/clspv/third_party/llvm/llvm/utils/unittest/googlemock/include -fdiagnostics-color -g   -DCLSPV_ONLINE_COMPILER -std=gnu++14 -MD -MT tests/api/CMakeFiles/api_tests.dir/images.cpp.o -MF tests/api/CMakeFiles/api_tests.dir/images.cpp.o.d -o tests/api/CMakeFiles/api_tests.dir/images.cpp.o -c ../tests/api/images.cpp
In file included from ../tests/api/images.cpp:18:
In file included from ../tests/../src/utils.hpp:17:
../tests/../src/log.hpp:20:10: fatal error: 'vulkan/vulkan.h' file not found
#include <vulkan/vulkan.h>
         ^~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by rjodinchr** [comment]
> CI seems to have an issue finding the vulkan headers
> 
> ```
> /bin/clang-8.0.0/bin/clang++  -DCOMPILER_AVAILABLE -I../external/OpenCL-Headers -I/tmpfs/src/github/clspv/third_party/llvm/llvm/include -Iexternal/clspv/third_party/llvm/include -I/tmpfs/src/github/clspv/third_party/llvm/llvm/utils/unittest/googletest/include -I../tests/../src -I../external/SPIRV-Tools/include -I../external/SPIRV-Headers/include -I/tmpfs/src/github/clspv/include -Iexternal/clspv/include -I/tmpfs/src/github/clspv/third_party/llvm/llvm/utils/unittest/googlemock/include -fdiagnostics-color -g   -DCLSPV_ONLINE_COMPILER -std=gnu++14 -MD -MT tests/api/CMakeFiles/api_tests.dir/images.cpp.o -MF tests/api/CMakeFiles/api_tests.dir/images.cpp.o.d -o tests/api/CMakeFiles/api_tests.dir/images.cpp.o -c ../tests/api/images.cpp
> In file included from ../tests/api/images.cpp:18:
> In file included from ../tests/../src/utils.hpp:17:
> ../tests/../src/log.hpp:20:10: fatal error: 'vulkan/vulkan.h' file not found
> #include <vulkan/vulkan.h>
>          ^~~~~~~~~~~~~~~~~
> 1 error generated.
> ```
#898

File: include/clspv/Option.h [hunk 1]
```diff
@@ -16,10 +16,32 @@
...
+}
+
 namespace Option {
 
+// Custom parser for parsing a comma separated list of builtin functions into a
+// std::set containing the equivalent list of BuiltinType enums. We use a
+// custom parser with a regular cl::opt instead of a cl::list because cl::list
+// requires you to list all possible choices (in this case all builtin
+// functions) and then includes all those choices in the helptext for the flag.
+struct BuiltinListParser
```
Comment (alan-baker) [inline include/clspv/Option.h:36]: This seems unnecessary in the header. Can remove the command line header include too.

**Review by alan-baker — CHANGES_REQUESTED** [review]
Looks good for the most part, but I'm not sure why the option parser is part of the public API.

**Comment by aarongreig** [comment]
thanks for the review

**Review by alan-baker — APPROVED** [review]

**Pull request merged**